### ISSUE TITLE
globally activated sms auth not set for new users

### ISSUE THREAD

**Issue by frisi** [collective/collective.smsauthenticator#15]
currently sms authentication is not activated for new users.

to reproduce:
* log in as site-admin
* install smsauthenticator
* enable "globally enabled" in @@sms-authenticator-settings
* create a new user (check send email to set password)
* optional: set a mobile number for this user
* open different browser
* open link and set password for newly created user
* login with new password

currently the user will be successfully logged in

what should happen is:
A) if a mobile number has been set: the user should be prompted for the code he/she received on their mobile phone or
B) (if no mobile number has been set) the user should be asked to set and verify his/her mobile number

the problem is, that for new users the checkbox `enable_two_step_verification` in `/@@user-information?userid=newuser` is not checked.

suggested solution: 
register event listener for user added event and set `enable_two_step_verification` if `globally_enabled == True`

**Comment by barseghyanartur** [comment]
@frisi:

What you describe is how [it was supposed to work and worked in past](https://pypi.python.org/pypi/collective.smsauthenticator#globally-enabled). If it now acts differently, something was likely accidentally broken.

**Comment by puittenbroek** [comment]
Yes, I changed this part because there is/should be a difference between 'enabling' (ie. People *can* use it) and "forcing" it. At the moment, there's no distinction.

See line here, upon 'first (global) enable' it's applied to all users.
https://github.com/collective/collective.smsauthenticator/blob/master/src/collective/smsauthenticator/browser/controlpanel.py#L154

**Comment by frisi** [comment]
@puittenbroek: about the distinction - i understood it this way:

* if you want to enable (ie. people can use it) smsauthentication: install this package
* if you want to force it (people - including new users - have to use it), check globally_enabled.


your change in the controlpanel.py only changed one thing:
existing users that do not have 2-factor-auth enabled, won't get in enabled just because an admin saves changes in @@sms-authenticator-settings (ie. ip whitelist) and did not change the globally_enabled settings.
i'm fine with that.


the change in the behaviour for new users is due to the commit that allowed to change the mobile number in the personal prefs (https://github.com/collective/collective.smsauthenticator/commit/4b9c44877fda0ad25ebf2cf5aad19647221070f8):

* IEnhancedUserDataSchema sets `enable_two_step_verification` to False
* and now the adapter writes this value into the memberdata properties.

unfortunatley, i think we can't change `False` to a value depending on the registry setting for globally_enabled, can we (@puittenbroek)?



to be on the save side we need to add a check if a user has permission to change 2-factor settings if globally_enabled is set to True

and change the implementation in the adapter:

```python

    def set_enable_two_step_verification(self, value):
        # warning: pseudo code!
        can_edit = checkpermission('collective.smsauthenticator: Toggle 2-Factor authentication')
        force_auth = api.registry.get('smsauthenticator.globally_enabled', False)
        
        if force_auth and not can_edit:
            # enabled by default, user must not change
            return
        elif force_auth and can_edit:
            # user has permission to change value althoug 2-factor is required
            return self.context.setMemberProperties({'enable_two_step_verification': value})
        
        # we don't force 2-factor
        return self.context.setMemberProperties({'enable_two_step_verification': value})
```

a site admin that creates a new user will have the required permission, the userdataschema sets to false and we still end up with a new user with 2-factor-auth disabled


if we do not find a way to make the default value for  IEnhancedUserDataSchema.enable_two_step_verification depend on globally_enabled, 
i think we need a event handler that sets enable_two_step_verification to True for newly created users in case globally_enabled == True


also a check for set_mobile number that checks the `collective.smsauthenticator: set mobile number w/o verification` permission
would make sense, what do you think?



to sum up:

* permission for changing moblie number without validation: site-admin by default
* permission to change 2-factor settings (even if globally_enabled == True)
  (defaults to manager and site-admin by default, integrators can grant it to member role if they want all users to be able to disable this setting)
* eventhandler that sets 2-factor settings for newly created user


@puittenbroek and @barseghyanartur 
* do you agree that these new permission make sense?
* do you know if we can change the default value in the schema depending on the registry setting
  if not - do you agree with the event-handler solution, or do you have a better idea?

**Comment by puittenbroek** [comment]
> if you want to enable (ie. people can use it) smsauthentication: install this package
> if you want to force it (people - including new users - have to use it), check globally_enabled.

That makes sense the way you describe it. I saw the 'globally enabled' as a master switch to allow users to enable/disable two factor. But indeed the installation of the product *is* that switch. The sms-auth settings "globally enabled" is the "force two factor" :)

> Eventhandler that sets 2-factor settings for newly created user

The eventhandler you mention already exists (even if a bit hidden), see here:
https://github.com/collective/collective.smsauthenticator/blob/master/src/collective/smsauthenticator/userdataschema.py#L118



> do you agree that these new permission make sense?

Yes, the new permissions make sense. That allows the use-case for which I intended the change without accidently allowing users to bypass the 'force auth'

> do you know if we can change the default value in the schema depending on the registry setting if not - do you agree with the event-handler solution, or do you have a better idea?

This is already present in the mentioned event listener, could it be that this is malfunctioning but error being swallowed? Most likely due to a permission problem. Perhaps we have to adopt the manager role in this part (via plone.api .. which we already use anyway).

**Comment by puittenbroek** [comment]
Note: perhaps change the name of the "globally enabled" settings to "force two factor".. or something similar to that. I work on the product and even I get confused ;) @barseghyanartur ?

**Comment by barseghyanartur** [comment]
@puittenbroek:

Force two factor seems fine to me.

**Comment by frisi** [comment]
would you
* A) simply change the title of the field `ISMSAuthenticatorSettings.globally_enabled` or
* B) are you considering to change the field name to `ISMSAuthenticatorSettings.force_2factor` (resulting in a backward incompatible change and requiring an upgrade step)?

**Comment by barseghyanartur** [comment]
I would rather change the title and leave the comment in the definition of the field.

**Comment by frisi** [comment]
luckily there is a way to compute the value of a zope.schema field at the time the addform is loaded:
http://docs.plone.org/external/plone.app.dexterity/docs/advanced/defaults.html
(so no need to fix/add the eventhandler.)

all we need to do is:

* [ ] rename title of "globally enabled"  to "Force two factor"
* [ ] use defaultFactory to check registry.globally_enabled for default value of IEnhancedUserDataSchema.enable_two_step_verification
* [ ] add permissions for editing mobile number of chaning enable_two_step_verfication as oulined at the end of comment https://github.com/collective/collective.smsauthenticator/issues/15#issuecomment-218481572

**Comment by tkimnguyen** [comment]
closing this since the PR was merged a while ago
